src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -268,59 +265,74 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
+            public ParameterLookupEntry(JsonPropertyInfo jsonPropertyInfo)
+            {
+                JsonPropertyInfo = jsonPropertyInfo;
+            }
+
+            public JsonPropertyInfo JsonPropertyInfo { get; }
+            public JsonPropertyInfo? DuplicateJsonPropertyInfo { get; set; } = null!;
+        }
+
+        private void InitializeConstructorParameters(ConstructorInfo constructorInfo)
         {
             ParameterInfo[] parameters = constructorInfo!.GetParameters();
-            Dictionary<string, JsonParameterInfo> parameterCache = new Dictionary<string, JsonParameterInfo>(
+            var parameterCache = new Dictionary<string, JsonParameterInfo>(
                 parameters.Length, Options.PropertyNameCaseInsensitive ? StringComparer.OrdinalIgnoreCase : null);
 
-            foreach (ParameterInfo parameterInfo in parameters)
+            // Cache the lookup from object property name to JsonPropertyInfo using a case-insensitive comparer.
+            // Case-insensitive is used to support both camel-cased parameter names and exact matches when C#
+            // record types or anonymous types are used.
+            // The property name key does not use [JsonPropertyName] or PropertyNamingPolicy since we only bind
+            // the parameter name to the object property name and do not use the JSON version of the name here.
+            var nameLookup = new Dictionary<string, ParameterLookupEntry>(PropertyCacheArray!.Length, StringComparer.OrdinalIgnoreCase);
+            foreach (JsonPropertyInfo jsonProperty in PropertyCacheArray!)
             {
-                MemberInfo? firstMatch = null;
-                bool isBound = false;
-
-                foreach (JsonPropertyInfo jsonPropertyInfo in PropertyCacheArray!)
+                string propertyName = jsonProperty.MemberInfo!.Name;
+                var entry = new ParameterLookupEntry(jsonProperty);
+                if (!JsonHelpers.TryAdd(nameLookup, propertyName, entry))
                 {
-                    // This is not null because it is an actual
-                    // property on a type, not a "policy property".
-                    MemberInfo memberInfo = jsonPropertyInfo.MemberInfo!;
-                    Debug.Assert(memberInfo is PropertyInfo || memberInfo is FieldInfo);
-
-                    Type memberType = memberInfo is PropertyInfo propertyInfo
-                        ? propertyInfo.PropertyType
-                        : Unsafe.As<FieldInfo>(memberInfo).FieldType;
+                    // More than one property has the same case-insensitive name.
+                    // Remember so we can throw a nice exception if this property is used as a parameter name.
+                    nameLookup[propertyName].DuplicateJsonPropertyInfo = jsonProperty;
+                }
+            }
 
-                    string camelCasePropName = JsonNamingPolicy.CamelCase.ConvertName(memberInfo.Name);
+            Type GetMemberType(MemberInfo memberInfo)
```

Comment:
```suggestion
            static Type GetMemberType(MemberInfo memberInfo)
```

Reply:
Thanks I'll change it afterwards as to not reset CI (unless something else comes up).